File: backend/src/schemas/members.py
"""Member request/response models including the joined Google state."""
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

GoogleStateName = Literal[
    "connected", "reconnect_needed", "revoked", "not_connected"
]


class GoogleState(BaseModel):
    status: GoogleStateName
    email: str | None = None
    connected_at: datetime | None = None


class MemberCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    nickname: str | None = None
    color: str = Field(min_length=1, max_length=32)


class MemberUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    nickname: str | None = None
    color: str | None = Field(default=None, min_length=1, max_length=32)


class MemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    family_id: UUID
    name: str
    nickname: str | None = None
    color: str
    status: Literal["active", "inactive"]
    is_setup_owner: bool
    google: GoogleState
    created_at: datetime

File: frontend/src/lib/api/members.ts
import { http, jsonBody } from "./http";
import type { MemberColor } from "@/components/fridge/types";

export type MemberStatus = "active" | "inactive";
export type GoogleStatus =
  | "connected"
  | "reconnect_needed"
  | "revoked"
  | "not_connected";

export interface GoogleStateResponse {
  status: GoogleStatus;
  email: string | null;
  connected_at: string | null;
}

export interface MemberResponse {
  id: string;
  family_id: string;
  name: string;
  nickname: string | null;
  color: MemberColor;
  status: MemberStatus;
  is_setup_owner: boolean;
  google: GoogleStateResponse;
  created_at: string;
}

export interface MemberCreateRequest {
  name: string;
  nickname?: string | null;
  color: MemberColor;
}

export type MemberUpdateRequest = Partial<MemberCreateRequest>;

export type MemberStatusFilter = "active" | "inactive" | "all";

export const membersApi = {
  list: (status: MemberStatusFilter = "active") =>
    http<MemberResponse[]>(`/members?status=${status}`),
  get: (id: string) => http<MemberResponse>(`/members/${id}`),
  create: (body: MemberCreateRequest) =>
    http<MemberResponse>("/members", {
      method: "POST",
      body: jsonBody(body),
    }),
  update: (id: string, body: MemberUpdateRequest) =>
    http<MemberResponse>(`/members/${id}`, {
      method: "PATCH",
      body: jsonBody(body),
    }),
  setActive: (id: string) =>
    http<MemberResponse>(`/members/${id}/set-active`, { method: "POST" }),
  setInactive: (id: string) =>
    http<MemberResponse>(`/members/${id}/set-inactive`, { method: "POST" }),
};
